src/Utilities/TimerManager.h

```diff
@@ -136,14 +137,11 @@ class TimerManager
 
 extern template class TimerManager<NewTimer>;
 extern template class TimerManager<FakeTimer>;
-
-extern TimerManager<NewTimer> timer_manager;
+inline static TimerManager<NewTimer> timer_manager;
```

Comment:
timer_manager is global. Using static makes one local copy in each cpp file that includes this .h. Right?

Reply:
```
(gdb) run "det_qmc_short_vmc_dmc_mwalkers.in.xml"
Starting program: /home/svh/Documents/qmcpack/build/bin/qmcpack "det_qmc_short_vmc_dmc_mwalkers.in.xml"
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff063165e in std::_Rb_tree_decrement(std::_Rb_tree_node_base*) () from /lib/x86_64-linux-gnu/libstdc++.so.6
(gdb) bt
#0  0x00007ffff063165e in std::_Rb_tree_decrement(std::_Rb_tree_node_base*) () from /lib/x86_64-linux-gnu/libstdc++.so.6
#1  0x0000555556640167 in std::_Rb_tree_iterator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >::operator-- (this=<synthetic pointer>) at /usr/include/c++/9/bits/stl_tree.h:302
#2  std::_Rb_tree<unsigned char, std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned char>, std::allocator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_get_insert_unique_pos (this=this@entry=0x555556831380 <qmcplusplus::timer_manager+96>, __k=@0x55555686ec20: 0 '\000')
    at /usr/include/c++/9/bits/stl_tree.h:2109
#3  0x000055555664024b in std::_Rb_tree<unsigned char, std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned char>, std::allocator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_get_insert_hint_unique_pos (__position=..., __k=@0x55555686ec20: 0 '\000', 
    this=0x555556831380 <qmcplusplus::timer_manager+96>) at /usr/include/c++/9/bits/stl_tree.h:2206
#4  std::_Rb_tree<unsigned char, std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned char>, std::allocator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_get_insert_hint_unique_pos (this=0x555556831380 <qmcplusplus::timer_manager+96>, __position=..., 
    __k=@0x55555686ec20: 0 '\000') at /usr/include/c++/9/bits/stl_tree.h:2192
#5  0x000055555663dc8d in std::_Rb_tree<unsigned char, std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::_Select1st<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::less<unsigned char>, std::allocator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::_M_emplace_hint_unique<std::piecewise_construct_t const&, std::tuple<unsigned char&&>, std::tuple<> >(std::_Rb_tree_const_iterator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, std::piecewise_construct_t const&, std::tuple<unsigned char&&>&&, std::tuple<>&&) (
    this=this@entry=0x555556831380 <qmcplusplus::timer_manager+96>, __pos=__pos@entry={first = 0 '\000', second = ""}) at /usr/include/c++/9/bits/stl_function.h:1155
#6  0x0000555556640e31 in std::map<unsigned char, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::less<unsigned char>, std::allocator<std::pair<unsigned char const, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > >::operator[] (__k=@0x7fffffffcb27: 0 '\000', this=0x555556831380 <qmcplusplus::timer_manager+96>) at /usr/include/c++/9/bits/stl_tree.h:348
#7  qmcplusplus::TimerManager<qmcplusplus::TimerType<std::chrono::_V2::system_clock> >::initializeTimer (this=this@entry=0x555556831320 <qmcplusplus::timer_manager>, t=...) at /home/svh/Documents/qmcpack/src/Utilities/TimerManager.cpp:51
#8  0x0000555556643287 in qmcplusplus::TimerManager<qmcplusplus::TimerType<std::chrono::_V2::system_clock> >::createTimer (this=this@entry=0x555556831320 <qmcplusplus::timer_manager>, myname="Block", mytimer=mytimer@entry=qmcplusplus::timer_level_coarse)
    at /home/svh/Documents/qmcpack/src/Utilities/TimerManager.cpp:76
#9  0x00005555562db469 in qmcplusplus::TimerList<qmcplusplus::TimerType<std::chrono::_V2::system_clock> >::TimerList<qmcplusplus::AFQMCTimerIDs> (this=0x5555568306e0 <qmcplusplus::AFQMCTimers>, timer_list=std::vector of length 23, capacity 23 = {...}, 
    timer_level=qmcplusplus::timer_level_coarse, manager=0x555556831320 <qmcplusplus::timer_manager>) at /home/svh/Documents/qmcpack/src/Utilities/TimerManager.h:169
#10 0x000055555568f29c in __static_initialization_and_destruction_0 (__priority=65535, __initialize_p=1) at /home/svh/Documents/qmcpack/src/AFQMC/AFQMCFactory.cpp:59
#11 0x000055555667074d in __libc_csu_init ()
#12 0x00007ffff033c010 in __libc_start_main (main=0x555555649a80 <main(int, char**)>, argc=2, argv=0x7fffffffd168, init=0x555556670700 <__libc_csu_init>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7fffffffd158) at ../csu/libc-start.c:264
#13 0x0000555555696aee in _start () at /usr/include/c++/9/bits/move.h:74
(gdb) 
```

Reply:
https://isocpp.org/wiki/faq/ctors#static-init-order-on-first-use

Reply:
yep. It is an initialization order issue of timer_manager and AFQMCTimers. AFQMC should have its own global timer_manager.

Reply:
Changed my mind. Making AFQMC its own global timer_manager actually makes timer printing at the end of QMCPACK run to print nothing. I think we should not expose the global timer manager in the header but expose a function getGlobalTimerManager() which does initialization of timer_manager on the first call.
Also timer_manager.createTimer is the most used way of calling timer_manager, I'm thinking if we can just add a syntactic sugar. Let me give it a quick try.

Reply:
Put up https://github.com/quantumsteve/qmcpack/pull/1